lib/rpmi_service_group_cppc.c

```diff
@@ -0,0 +1,808 @@
+			   const rpmi_uint8_t *request_data,
+			   rpmi_uint16_t *response_datalen,
+			   rpmi_uint8_t *response_data)
+{
+	enum rpmi_error status;
+	rpmi_uint32_t start_index, max_entries, hart_count, hart_id;
+	struct rpmi_cppc_group *cppcgrp = group->priv;
+	rpmi_uint32_t *resp = (void *)response_data;
+	rpmi_uint32_t returned, remaining, i;
+
+	hart_count = rpmi_hsm_hart_count(cppcgrp->hsm);
+	max_entries = RPMI_MSG_DATA_SIZE(trans->slot_size) - (3 * sizeof(*resp));
+	max_entries = rpmi_env_div32(max_entries, sizeof(*resp));
+
+	start_index = rpmi_to_xe32(trans->is_be, ((const rpmi_uint32_t *)request_data)[0]);
+
+	if (start_index <= hart_count) {
+		returned = max_entries < (hart_count - start_index) ?
+			max_entries : (hart_count - start_index);
+		for (i = 0; i < returned; i++) {
+			hart_id = rpmi_hsm_hart_index2id(cppcgrp->hsm, start_index + i);
+			resp[3 + i] = rpmi_to_xe32(trans->is_be, hart_id);
+		}
+		remaining = hart_count - (start_index + returned);
+		status = RPMI_SUCCESS;
+	} else {
+		returned = 0;
+		remaining = hart_count;
+		status = RPMI_ERR_INVAL;
+	}
+
+	*response_datalen = (returned + 3) * sizeof(*resp);
+
+	resp[0] = rpmi_to_xe32(trans->is_be, (rpmi_uint32_t)status);
+	resp[1] = rpmi_to_xe32(trans->is_be, remaining);
+	resp[2] = rpmi_to_xe32(trans->is_be, returned);
+
+	return RPMI_SUCCESS;
+}
+
+static struct rpmi_service rpmi_cppc_services[RPMI_CPPC_SRV_ID_MAX] = {
+	[RPMI_CLK_SRV_ENABLE_NOTIFICATION] = {
+		.service_id = RPMI_CLK_SRV_ENABLE_NOTIFICATION,
+		.min_a2p_request_datalen = 4,
+		.process_a2p_request = NULL,
+	},
+	[RPMI_CPPC_SRV_PROBE_REG] = {
+		.service_id = RPMI_CPPC_SRV_PROBE_REG,
+		.min_a2p_request_datalen = 8,
+		.process_a2p_request = rpmi_cppc_sg_probe_reg,
+	},
+	[RPMI_CPPC_SRV_READ_REG] = {
+		.service_id = RPMI_CPPC_SRV_READ_REG,
+		.min_a2p_request_datalen = 8,
+		.process_a2p_request = rpmi_cppc_sg_read_reg,
+	},
+	[RPMI_CPPC_SRV_WRITE_REG] = {
+		.service_id = RPMI_CPPC_SRV_WRITE_REG,
+		.min_a2p_request_datalen = 16,
+		.process_a2p_request = rpmi_cppc_sg_write_reg,
+	},
+	[RPMI_CPPC_SRV_GET_FAST_CHANNEL_REGION] = {
+		.service_id = RPMI_CPPC_SRV_GET_FAST_CHANNEL_REGION,
+		.min_a2p_request_datalen = 0,
+		.process_a2p_request = rpmi_cppc_sg_get_fast_channel_region,
+	},
+	[RPMI_CPPC_SRV_GET_FAST_CHANNEL_OFFSET] = {
+		.service_id = RPMI_CPPC_SRV_GET_FAST_CHANNEL_OFFSET,
+		.min_a2p_request_datalen = 4,
+		.process_a2p_request = rpmi_cppc_sg_get_fast_channel_offset,
+	},
+	[RPMI_CPPC_SRV_GET_HART_LIST] = {
+		.service_id = RPMI_CPPC_SRV_GET_HART_LIST,
+		.min_a2p_request_datalen = 4,
+		.process_a2p_request = rpmi_cppc_sg_get_hart_list,
+	},
+};
+
+static enum rpmi_error rpmi_cppc_process_events(struct rpmi_service_group *group)
+{
+
+	enum rpmi_error status = RPMI_SUCCESS;
+	rpmi_uint32_t hart_idx, desired_perf;
+	union rpmi_cppc_hart_fastchan *fc_hart;
+	struct rpmi_cppc_group *cppcgrp = group->priv;
+
+	for (hart_idx = 0; hart_idx < cppcgrp->hart_count; hart_idx++) {
```

Comment:
Need to check CPPC mode ?

Reply:
In this version i just wanted to introduce that librpmi will have CPPC modes. But since even the linux cppc cpufreq driver does not implement this. So didnt use mode checking anywhere in code.

Reply:
I understand that but since this implementation does not have active mode, we should have a check to ensure that we only process fast channels for passive mode.